e2e/test/AppContextTests.cs

```diff
@@ -0,0 +1,332 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Client.Common;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    [TestClass]
+    [TestCategory("IoTHub-E2E")]
+    [TestCategory("AppContext")]
+    public class AppContextTests
+    {
+        private readonly string DevicePrefix = $"E2E_{nameof(AppContextTests)}_";
+        private static TestLogging _log = TestLogging.GetInstance();
+
+        private readonly ConsoleEventListener _listener;
+
+        public AppContextTests()
+        {
+#if !NET451
+            AppContext.SetSwitch(AppContextConstants.DisableObjectDisposedExceptionForReceiveAsync, true);
+#endif
+
+            _listener = TestConfig.StartEventListener();
+        }
+
+        #region ReceiveAsyncAfterDisposeAsync
+
+        [TestMethod]
+        public async Task AppConfig_ReceiveAsyncAfterDispose_Sasl_Amqp()
+        {
+
+            bool isFlagSet = false;
+            bool flag = false;
+
+#if !NET451
+            isFlagSet = AppContext.TryGetSwitch(AppContextConstants.DisableObjectDisposedExceptionForReceiveAsync, out flag);
+#endif
+
+            if (isFlagSet)
+            {
+                await ReceiveAsyncAfterDisposeAsync(TestDeviceType.Sasl, Client.TransportType.Amqp_Tcp_Only, false).ConfigureAwait(false);
+            }
+            else
+            {
+                Assert.Fail("AppContext flag DisableObjectDisposedExceptionForReceiveAsync not set");
+            }
+        }
+
+        [TestMethod]
+        public async Task AppConfig_ReceiveAsyncWithCancellationTokenAfterDispose_Sasl_Amqp()
+        {
+
+            bool isFlagSet = false;
+            bool flag = false;
+
+#if !NET451
+            isFlagSet = AppContext.TryGetSwitch(AppContextConstants.DisableObjectDisposedExceptionForReceiveAsync, out flag);
+#endif
+
+            if (isFlagSet)
+            {
+                await ReceiveAsyncAfterDisposeAsync(TestDeviceType.Sasl, Client.TransportType.Amqp_Tcp_Only, true).ConfigureAwait(false);
+            }
+            else
+            {
+                Assert.Fail("AppContext flag DisableObjectDisposedExceptionForReceiveAsync not set");
+            }
+        }
+
+        [Ignore("ReceiveAsync hangs due to bug in DotNetty (https://github.com/Azure/DotNetty/issues/502)")]
```

Comment:
We should tag this to an issue in our repo as well, so that we can identify and re-enable these tests once the dotnetty issue is resolved.

Reply:
+1 on that - we should not have ignored tests without a tracking issue in our repo.